File: com.trollworks.gcs/src/com/trollworks/gcs/criteria/StringCompareType.java
/*
 * Copyright ©1998-2020 by Richard A. Wilkes. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, version 2.0. If a copy of the MPL was not distributed with
 * this file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * This Source Code Form is "Incompatible With Secondary Licenses", as
 * defined by the Mozilla Public License, version 2.0.
 */

package com.trollworks.gcs.criteria;

import com.trollworks.gcs.utility.I18n;

/** The allowed string comparison types. */
public enum StringCompareType {
    /** The comparison for "is anything". */
    ANY {
        @Override
        public String toString() {
            return I18n.Text("is anything");
        }

        @Override
        public boolean matches(String qualifier, String data) {
            return true;
        }
    },
    /** The comparison for "is". */
    IS {
        @Override
        public String toString() {
            return I18n.Text("is");
        }

        @Override
        public boolean matches(String qualifier, String data) {
            return data.equalsIgnoreCase(qualifier);
        }
    },
    /** The comparison for "is not". */
    IS_NOT {
        @Override
        public String toString() {
            return I18n.Text("is not");
        }

        @Override
        public boolean matches(String qualifier, String data) {
            return !data.equalsIgnoreCase(qualifier);
        }
    },
    /** The comparison for "contains". */
    CONTAINS {
        @Override
        public String toString() {
            return I18n.Text("contains");
        }

        @Override
        public boolean matches(String qualifier, String data) {
            return data.toLowerCase().contains(qualifier.toLowerCase());
        }
    },
    /** The comparison for "does not contain". */
    DOES_NOT_CONTAIN {
        @Override
        public String toString() {
            return I18n.Text("does not contain");
        }

        @Override
        public boolean matches(String qualifier, String data) {
            return !data.toLowerCase().contains(qualifier.toLowerCase());
        }
    },
    /** The comparison for "starts with". */
    STARTS_WITH {
        @Override
        public String toString() {
            return I18n.Text("starts with");
        }

        @Override
        public boolean matches(String qualifier, String data) {
            return data.toLowerCase().startsWith(qualifier.toLowerCase());
        }
    },
    /** The comparison for "does not start with". */
    DOES_NOT_START_WITH {
        @Override
        public String toString() {
            return I18n.Text("does not start with");
        }

        @Override
        public boolean matches(String qualifier, String data) {
            return !data.toLowerCase().startsWith(qualifier.toLowerCase());
        }
    },
    /** The comparison for "ends with". */
    ENDS_WITH {
        @Override
        public String toString() {
            return I18n.Text("ends with");
        }

        @Override
        public boolean matches(String qualifier, String data) {
            return data.toLowerCase().endsWith(qualifier.toLowerCase());
        }
    },
    /** The comparison for "does not end with". */
    DOES_NOT_END_WITH {
        @Override
        public String toString() {
            return I18n.Text("does not end with");
        }

        @Override
        public boolean matches(String qualifier, String data) {
            return !data.toLowerCase().endsWith(qualifier.toLowerCase());
        }
    };

    /**
     * @param qualifier The qualifier.
     * @return The description of this comparison type.
     */
    public String describe(String qualifier) {
        return toString() + " \"" + qualifier + '"';
    }

    /**
     * Performs a comparison.
     *
     * @param qualifier The qualifier to use in conjunction with this {@link StringCompareType}.
     * @param data      The data to check.
     * @return Whether the data matches the criteria or not.
     */
    public abstract boolean matches(String qualifier, String data);
}

File: model/criteria/string_compare_type.go
/*
 * Copyright ©1998-2022 by Richard A. Wilkes. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, version 2.0. If a copy of the MPL was not distributed with
 * this file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * This Source Code Form is "Incompatible With Secondary Licenses", as
 * defined by the Mozilla Public License, version 2.0.
 */

package criteria

import (
	"strings"

	"github.com/richardwilkes/toolbox/i18n"
)

// Possible StringCompareType values.
const (
	Any              = StringCompareType("")
	Is               = StringCompareType("is")
	IsNot            = StringCompareType("is_not")
	Contains         = StringCompareType("contains")
	DoesNotContain   = StringCompareType("does_not_contain")
	StartsWith       = StringCompareType("starts_with")
	DoesNotStartWith = StringCompareType("does_not_start_with")
	EndsWith         = StringCompareType("ends_with")
	DoesNotEndWith   = StringCompareType("does_not_end_with")
)

// AllStringCompareTypes is the complete set of StringCompareType values.
var AllStringCompareTypes = []StringCompareType{
	Any,
	Is,
	IsNot,
	Contains,
	DoesNotContain,
	StartsWith,
	DoesNotStartWith,
	EndsWith,
	DoesNotEndWith,
}

// StringCompareType holds the type for a string comparison.
type StringCompareType string

// EnsureValid ensures this is of a known value.
func (s StringCompareType) EnsureValid() StringCompareType {
	for _, one := range AllStringCompareTypes {
		if one == s {
			return s
		}
	}
	return AllStringCompareTypes[0]
}

// String implements fmt.Stringer.
func (s StringCompareType) String() string {
	switch s {
	case Any:
		return i18n.Text("is anything")
	case Is:
		return i18n.Text("is")
	case IsNot:
		return i18n.Text("is not")
	case Contains:
		return i18n.Text("contains")
	case DoesNotContain:
		return i18n.Text("does not contain")
	case StartsWith:
		return i18n.Text("starts with")
	case DoesNotStartWith:
		return i18n.Text("does not start with")
	case EndsWith:
		return i18n.Text("ends with")
	case DoesNotEndWith:
		return i18n.Text("does not end with")
	default:
		return Any.String()
	}
}

// Describe returns a description of this StringCompareType using a qualifier.
func (s StringCompareType) Describe(qualifier string) string {
	v := s.EnsureValid()
	if v == Any {
		return v.String()
	}
	return v.String() + ` "` + qualifier + `"`
}

// Matches performs a comparison and returns true if the data matches.
func (s StringCompareType) Matches(qualifier, data string) bool {
	switch s {
	case Any:
		return true
	case Is:
		return strings.EqualFold(data, qualifier)
	case IsNot:
		return !strings.EqualFold(data, qualifier)
	case Contains:
		return strings.Contains(strings.ToLower(data), strings.ToLower(qualifier))
	case DoesNotContain:
		return !strings.Contains(strings.ToLower(data), strings.ToLower(qualifier))
	case StartsWith:
		return strings.HasPrefix(strings.ToLower(data), strings.ToLower(qualifier))
	case DoesNotStartWith:
		return !strings.HasPrefix(strings.ToLower(data), strings.ToLower(qualifier))
	case EndsWith:
		return strings.HasSuffix(strings.ToLower(data), strings.ToLower(qualifier))
	case DoesNotEndWith:
		return !strings.HasSuffix(strings.ToLower(data), strings.ToLower(qualifier))
	default:
		return Any.Matches(qualifier, data)
	}
}

// ExtractStringCompareTypeIndex extracts the index from a string.
func ExtractStringCompareTypeIndex(str string) int {
	for i, one := range AllStringCompareTypes {
		if strings.EqualFold(string(one), str) {
			return i
		}
	}
	return 0
}

// PrefixedStringCompareTypeChoices returns the set of StringCompareType choices as strings with a prefix.
func PrefixedStringCompareTypeChoices(prefix string) []string {
	choices := make([]string, len(AllStringCompareTypes))
	for i, choice := range AllStringCompareTypes {
		choices[i] = prefix + " " + choice.String()
	}
	return choices
}
